server/gateway.go

```diff
@@ -1436,21 +1418,30 @@ func (s *Server) removeRemoteGatewayConnection(c *client) {
 	}
 	gw.Unlock()
 	s.removeFromTempClients(cid)
-}
 
-// This allows some cleanup with guarantee that readloop has
-// exited and so no protocol message is being processed.
-func (c *client) gatewayOutboundConnectionReadLoopExited() {
-	qSubsRemoved := int64(0)
-	c.mu.Lock()
-	for _, sub := range c.subs {
-		if sub.queue != nil {
-			qSubsRemoved++
+	if isOutbound {
+		var subsa [1024]*subscription
+		var subs = subsa[:0]
+
+		// Update number of totalQSubs for this gateway
+		qSubsRemoved := int64(0)
+		c.mu.Lock()
+		for _, sub := range c.subs {
+			if sub.queue != nil {
+				qSubsRemoved++
+			} else if sub.qw > 0 {
```

Comment:
This seems brittle.

Reply:
Agreed. Posted an update. I like the new code much better.